packages/extension-base/src/services/balance-service/transfer/ton-transfer.ts

```diff
@@ -0,0 +1,138 @@
+import { estimateTonTxFee } from '@subwallet/extension-base/services/balance-service/helpers/subscribe/ton/utils';
+import { _TonApi } from '@subwallet/extension-base/services/chain-service/types';
+import { _getContractAddressOfToken, _isJettonToken, _isNativeToken } from '@subwallet/extension-base/services/chain-service/utils';
+import { keyring } from '@subwallet/ui-keyring';
+import { beginCell, fromNano, internal, MessageRelaxed, toNano } from '@ton/core';
+import { Address, JettonMaster, JettonWallet, WalletContractV4 } from '@ton/ton';
+
+interface TonTransactionConfigProps {
+  tokenInfo: _ChainAsset;
+  from: string,
+  to: string,
+  networkKey: string,
+  value: string,
+  transferAll: boolean,
+  tonApi: _TonApi
+}
+
+export interface TonTransactionConfig {
+  from: string,
+  to: string,
+  networkKey: string,
+  value: string,
+  messages: MessageRelaxed[]
+  estimateFee: string;
+  seqno: number
+}
+
+export async function createTonTransaction ({ from, networkKey, to, tokenInfo, tonApi, transferAll, value }: TonTransactionConfigProps): Promise<[TonTransactionConfig | null, string]> {
+  if (_isNativeToken(tokenInfo)) {
+    return createTonNativeTransaction({ from, to, networkKey, tokenInfo, value: value || '0', transferAll: transferAll, tonApi });
+  }
+
+  if (_isJettonToken(tokenInfo)) {
+    return createJettonTransaction({ from, to, networkKey, tokenInfo, value: value || '0', transferAll: transferAll, tonApi });
+  }
+
+  return [null, value];
+}
+
+async function createTonNativeTransaction ({ from, networkKey, to, tonApi, transferAll, value }: TonTransactionConfigProps): Promise<[TonTransactionConfig | null, string]> {
+  const keyPair = keyring.getPair(from);
+  const tonAddress = Address.parse(from);
+  const walletContract = WalletContractV4.create({ workchain: WORKCHAIN, publicKey: Buffer.from(keyPair.publicKey) });
+  const contract = tonApi.open(walletContract);
+  const seqno = await contract.getSeqno();
+
+  let transferValue: string | undefined;
+
+  const messages = [
+    internal({
+      to: to,
+      value: fromNano(value),
+      bounce: false // todo: check and update the send bounced logic
+    })
+  ];
+
+  const estimateFee = await estimateTonTxFee(tonApi, messages, walletContract);
+
+  if (transferAll) {
+    const balance = await tonApi.getBalance(tonAddress);
+
+    transferValue = (balance - estimateFee).toString();
+  }
+
+  const transactionObject = {
+    from,
+    to,
+    networkKey,
+    value: transferValue ?? value,
+    messages,
+    estimateFee: estimateFee.toString(),
+    seqno
+  } as TonTransactionConfig;
+
+  return [transactionObject, transactionObject.value];
+}
+
+async function createJettonTransaction ({ from, networkKey, to, tokenInfo, tonApi, transferAll, value }: TonTransactionConfigProps): Promise<[TonTransactionConfig | null, string]> {
+  const keyPair = keyring.getPair(from);
```

Comment:
I don't think we should access keyring directly

Reply:
The pubKey have to be accessed to get create `wallet contract` - need for estimating tx fee.